File: cpp/src/csv_log.cpp
#include "bench/csv_log.hpp"

#include <cstdint>
#include <stdexcept>
#include <tuple>
#include <vector>

namespace bench {

CsvLogger::CsvLogger(const std::string& path) {
  f_ = std::fopen(path.c_str(), "w");
  if (!f_) throw std::runtime_error("cannot create CSV: " + path);
  std::fprintf(f_,
               "Language,StringOrStream,TestDataName,Repetitions,RepetitionIndex,SerializerName,"
               "SerializerVersion,TimeSer,TimeDeser,Size,TimeSerAndDeser,OpPerSecSer,OpPerSecDeser,"
               "OpPerSecSerAndDeser,MemoryPeakBytes,FidelityScore,NativeKind,StreamMode,"
               "DataTypeInstanceCount,TypeConfigHash\n");
}

CsvLogger::~CsvLogger() {
  if (f_) std::fclose(f_);
}

void CsvLogger::write_row(const std::string& mode, const std::string& test_data, int reps,
                          int rep_idx, const std::string& ser, const std::string& version,
                          uint64_t ser_ns, uint64_t deser_ns, size_t size, double fidelity,
                          const std::string& native_kind, const std::string& stream_mode,
                          int instance_count, const std::string& type_config_hash) {
  uint64_t tot = ser_ns + deser_ns;
  double ops_s = ser_ns ? 1e9 / static_cast<double>(ser_ns) : 0;
  double ops_d = deser_ns ? 1e9 / static_cast<double>(deser_ns) : 0;
  double ops_t = tot ? 1e9 / static_cast<double>(tot) : 0;
  if (instance_count < 1) instance_count = 1;
  std::fprintf(f_,
               "cpp,%s,%s,%d,%d,%s,%s,%llu,%llu,%zu,%llu,%.6f,%.6f,%.6f,0,%.1f,%s,%s,%d,%s\n",
               mode.c_str(), test_data.c_str(), reps, rep_idx, ser.c_str(), version.c_str(),
               static_cast<unsigned long long>(ser_ns), static_cast<unsigned long long>(deser_ns),
               size, static_cast<unsigned long long>(tot), ops_s, ops_d, ops_t, fidelity,
               native_kind.c_str(), stream_mode.c_str(), instance_count, type_config_hash.c_str());
}

void save_errors(
    const std::string& path,
    const std::vector<std::tuple<std::string, std::string, std::string, int, std::string>>& errors) {
  if (errors.empty()) {
    std::remove(path.c_str());
    return;
  }
  FILE* f = std::fopen(path.c_str(), "w");
  if (!f) return;
  std::fprintf(f, "TestDataName,SerializerName,StringOrStream,Repetition,ErrorText\n");
  for (const auto& e : errors) {
    std::string text = std::get<4>(e);
    for (char& c : text) {
      if (c == '\n' || c == '\r') c = ' ';
      if (c == ',') c = ';';
    }
    std::fprintf(f, "%s,%s,%s,%d,%s\n", std::get<0>(e).c_str(), std::get<1>(e).c_str(),
                 std::get<2>(e).c_str(), std::get<3>(e), text.c_str());
  }
  std::fclose(f);
}

}  // namespace bench

File: swift/Sources/SerializerBenchmarkCore/CsvLog.swift
import Foundation

/// RFC-style CSV field escape (quote when needed).
private func escapeCSV(_ field: String) -> String {
    if field.contains(",") || field.contains("\"") || field.contains("\n") || field.contains("\r") {
        return "\"\(field.replacingOccurrences(of: "\"", with: "\"\""))\""
    }
    return field
}

public final class CsvLogger {
    private let handle: FileHandle
    private let path: URL

    public init(path: URL) throws {
        self.path = path
        _ = FileManager.default.createFile(atPath: path.path, contents: nil)
        self.handle = try FileHandle(forWritingTo: path)
        let header = "Language,StringOrStream,TestDataName,Repetitions,RepetitionIndex,SerializerName,SerializerVersion,TimeSer,TimeDeser,Size,TimeSerAndDeser,OpPerSecSer,OpPerSecDeser,OpPerSecSerAndDeser,MemoryPeakBytes,FidelityScore,NativeKind,StreamMode,DataTypeInstanceCount,TypeConfigHash,RunOrder,SchedulePosition\n"
        handle.write(Data(header.utf8))
    }

    public func writeRow(
        mode: String,
        testDataName: String,
        repetitions: UInt32,
        repetitionIndex: UInt32,
        serializerName: String,
        serializerVersion: String,
        timeSer: UInt64,
        timeDeser: UInt64,
        size: Int,
        fidelity: Double,
        nativeKind: String,
        streamMode: String,
        instanceCount: Int,
        typeConfigHash: String,
        runOrder: Int = -1,
        schedulePosition: Int = -1
    ) {
        let total = timeSer &+ timeDeser
        let opsSer = timeSer > 0 ? 1e9 / Double(timeSer) : 0
        let opsDeser = timeDeser > 0 ? 1e9 / Double(timeDeser) : 0
        let opsTot = total > 0 ? 1e9 / Double(total) : 0
        let ro = runOrder >= 0 ? String(runOrder) : ""
        let sp = schedulePosition >= 0 ? String(schedulePosition) : ""
        let line = [
            escapeCSV("swift"),
            escapeCSV(mode),
            escapeCSV(testDataName),
            String(repetitions),
            String(repetitionIndex),
            escapeCSV(serializerName),
            escapeCSV(serializerVersion),
            String(timeSer),
            String(timeDeser),
            String(size),
            String(total),
            String(format: "%.6f", opsSer),
            String(format: "%.6f", opsDeser),
            String(format: "%.6f", opsTot),
            "0",
            String(format: "%.2f", fidelity),
            escapeCSV(nativeKind),
            escapeCSV(streamMode),
            String(instanceCount),
            escapeCSV(typeConfigHash),
            ro,
            sp,
        ].joined(separator: ",") + "\n"
        handle.write(Data(line.utf8))
    }

    public func close() {
        try? handle.close()
    }
}

public struct BenchRunError {
    public var testDataName: String
    public var serializerName: String
    public var stringOrStream: String
    public var repetition: UInt32
    public var errorText: String
}

public func saveErrors(path: URL, errors: [BenchRunError]) {
    if errors.isEmpty {
        try? FileManager.default.removeItem(at: path)
        return
    }
    var text = "TestDataName,SerializerName,StringOrStream,Repetition,ErrorText\n"
    for e in errors {
        text += [
            escapeCSV(e.testDataName),
            escapeCSV(e.serializerName),
            escapeCSV(e.stringOrStream),
            String(e.repetition),
            escapeCSV(e.errorText.replacingOccurrences(of: "\n", with: " ")),
        ].joined(separator: ",") + "\n"
    }
    try? text.write(to: path, atomically: true, encoding: .utf8)
}
